move ProtoReader from lib/events to api/sessionrecording as Reader

dustinspecker opened the pull request:
This pull request does the following:

- moves `ProtoReader` from `lib/events` to `api/sessionrecording` so that it may be imported by users to parse session recordings in Protobuf binary format
- renames `ProtoReader` to `Reader` to avoid leaking implementation detail in name

This pull request is focused on moving and renaming, while #50883 makes changes to implementation and the API.

**This pull request should not be merged until #50883 is also ready to be merged, since we do not want to support this API as-is.**

changelog: add api/sessionrecording.Reader to support parsing and reading session recordings

dustinspecker committed:
feat(api/sessionrecording): remove anything unrelated to ProtoReader

This commit removes unneeded code from api/sessionrecording and updates
the package name.

It removes the same code from lib/events and updates usages of this new
sessionrecording package.

dustinspecker committed:
refactor(sessionrecording): rename sessionlog file to gzip_reader

This file only contains gzipReader, so rename for clarity.

dustinspecker committed:
refactor(sessionrecording): rename stream file to session_recording

dustinspecker committed:
refactor(api/sessionrecording): rename ProtoReader to Reader

The name no longer leaks the implementation. Went with Reader over
SessionRecordingReader to prevent a stutter of writing
`sessionrecording.SessionRecordingReader`.

dustinspecker commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,287 @@
+)
+
+// Reader reads protobuf encoding from reader
+type Reader struct {
```
Open to other names. It was suggested to avoid `ProtoReader` since its name leaked the implementation. I had this as `SessionRecordingReader`, but then it reads as a stutter when using, e.g. `sessionrecording.SessionRecordingReader` versus `sessionrecording.Reader`.

zmb3 replied:
`Reader` is a great name, since callers will refer to it as `sessionrecording.Reader`.

I would avoid mentioning protobuf in the godoc though.

dustinspecker replied:
Good call, updated

dustinspecker commented on api/sessionrecording/session_recording_test.go:
```diff
@@ -0,0 +1,31 @@
+package sessionrecording_test
```
The tests are pretty light within `api/sessionrecording`, but looks like `lib/events` is testing most of `Reader`. I moved the only test case that seemed to be testing `Reader` directly. Everything else looks like it's using `Reader` to verify writes are as expected.

dustinspecker committed:
fix(lib/events): remove already moved unit test

TestReadCorruptedReading was moved to api/sessionrecording.

zmb3 commented on api/defaults/defaults.go:
```diff
@@ -139,6 +139,11 @@ func EnhancedEvents() []string {
 }
 
+const (
+	// MaxIterationLimit is max iteration limit
```
I'd probably just use a private const for this. It's too generic as-is to put here.

dustinspecker replied:
unexported and changed `lib/defaults` to also have a value of `1000`.

zmb3 commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,303 @@
+limitations under the License.
+*/
+
+package sessionrecording
```
Let's add the package-level godoc here.

```suggestion
// Package sessionrecording can be used to read Teleport's
// session recording files.
package sessionrecording
```

dustinspecker replied:
Thank you, added

zmb3 commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,303 @@
+
+const (
+	// Int32Size is a constant for 32 bit integer byte size
+	Int32Size = 4
```
Now that this is in API, we should try to export only what is absolutely necessary. If these values aren't needed outside the package, let's unexport them.

dustinspecker replied:
I moved the `*Header` constants back to `lib/events` since `api/sessionrecording` isn't using them. I was using them in another repository, but not super concerned about not being able to import them directly, especially since this package doesn't need them.


---

What are your thoughts on these remaining constants? `Int32Size`, `Int64Size`, `MaxProtoMessageSizeBytes`, and `ProtoStreamV1` are used by both `api` and `lib`.

1. Is it cool to keep these exported to keep them in sync between the two packages?
1. Or do you prefer to keep these unexported to avoid users from relying on these? And both `lib` and `api` can both have their own respective constants.

I'm good either way. I was leaning towards the first choice, but can understand not wanting users to rely on these constants.

zmb3 commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,303 @@
+	ProtoStreamV1RecordHeaderSize = Int32Size
+)
+
+// NewReader returns a new reader with slice pool
```
```suggestion
// NewReader returns a new session recording reader.
```
I wouldn't mention slice pool.

dustinspecker replied:
good call, removed

zmb3 commented on api/sessionrecording/session_recording_test.go:
Consider including a testable example: https://go.dev/blog/examples

Not sure if it makes sense without a real recording file, but would be nice to show users how leverage the reader.

dustinspecker replied:
I like this idea. I added an Example test. I created a new session file that is a not a real session, but has just enough metadata for the example to pass. e.g. it's not usable via `tsh play` or anything like that.

dustinspecker committed:
docs(api/sessionrecording): add package docs

dustinspecker committed:
docs(api/sessionrecording): remove references to protobuf or slice pool

dustinspecker committed:
docs(api/sessionrecording): add example for NewReader

dustinspecker committed:
fix(api/sessionrecording): unexport maxIterationLimit

This is simply a constant we use to break an infinite loop.

dustinspecker committed:
fix(api/sessionrecording): move back HeaderSize constants to lib/events

These constants are not used by api/sessionrecording itself, so do not
export.

dustinspecker committed:
fix(lib): use MaxProtoMessageSizeBytes from api/sessionrecording

dustinspecker commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,298 @@
+}
+
+// ToFields returns a copy of the stats to be used as log fields
+func (p ReaderStats) ToFields() map[string]any {
```
Thank you for that suggestion! I hadn't heard about `slog.LogValuer` before. Replaced `ToFields()` with `LogValue()`.

zmb3 commented on api/sessionrecording/gzip_reader.go:
```diff
@@ -0,0 +1,60 @@
+	"github.com/gravitational/trace"
+)
+
+// gzipReader wraps file, on close close both gzip writer and file
```
```suggestion
// gzipReader wraps an io.Reader, and closes both the gzip writer and underlying reader
```

I know this godoc existed before, but we might as well improve it while we're here.

There's no requirement for a "file" anywhere in this API.

dustinspecker replied:
Updated this godoc and then ended up removing this file entirely.

zmb3 commented on api/sessionrecording/gzip_reader.go:
```diff
@@ -0,0 +1,60 @@
+	inner io.ReadCloser
+}
+
+// Close closes file and gzip writer
```
```suggestion
```

I don't know that this godoc is even necessary.

dustinspecker replied:
Updated this godoc and then ended up removing this file entirely.

zmb3 commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,298 @@
+}
+
+const (
+	// protoReaderStateInit is ready to start reading the next part
+	protoReaderStateInit = 0
+	// protoReaderStateCurrent will read the data from the current part
+	protoReaderStateCurrent = iota
+	// protoReaderStateEOF indicates that reader has completed reading
+	// all parts
+	protoReaderStateEOF = iota
+	// protoReaderStateError indicates that reader has reached internal
+	// error and should close
+	protoReaderStateError = iota
```
```suggestion
	// protoReaderStateInit is ready to start reading the next part
	protoReaderStateInit = iota
	// protoReaderStateCurrent will read the data from the current part
	protoReaderStateCurrent
	// protoReaderStateEOF indicates that reader has completed reading
	// all parts
	protoReaderStateEOF
	// protoReaderStateError indicates that reader has reached internal
	// error and should close
	protoReaderStateError
```

dustinspecker replied:
updated - good catch

zmb3 commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,298 @@
+	TotalEvents int64
+}
+
+// ToFields returns a copy of the stats to be used as log fields
```
```suggestion
// LogValue returns a copy of the stats to be used as log fields
```

dustinspecker replied:
Thank you! I should have seen this. 🤦

zmb3 commented on api/sessionrecording/gzip_reader.go:
```diff
@@ -0,0 +1,60 @@
+	return trace.NewAggregate(errors...)
+}
+
+func newGzipReader(reader io.ReadCloser) (*gzipReader, error) {
```
On closer look, we made our own `gzipReader` instead of just using Go's `gzip.Reader` directly because we wanted to accept a `ReadCloser` and close it.

I wonder what it would look like if we shed our custom `gzipReader` completely and just had the `sessionrecording.Reader` keep a reference to the underlying reader to close it.

(Or even do what the gzip package does, and accept an `io.Reader` instead of an `io.ReadCloser`, and leave it up to the client to close things down if necessary)

dustinspecker replied:
Really great idea. Thank you for the thorough review!

I started with changing our `gzipReader` to only accept an `io.Reader` and realized sessionrecording was always using `io.NopCloser`, so that was an easy change. And then it was mostly straight forward to remove `gzipReader` entirely and now `sessionrecording.Reader` constructs a `gzip.Reader` with multistream disabled. I need to come back to this because I do not like the name `inner`.

Reviewing by commit instead of the whole diff might make the above easier to see.

zmb3 commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,298 @@
+	gzipReader   *gzipReader
+	padding      int64
+	reader       io.Reader
+	sizeBytes    [Int64Size]byte
+	messageBytes [MaxProtoMessageSizeBytes]byte
```
These two fields are only ever used in `Read`, so they could be extracted to local variables instead of fields on `Reader`.

dustinspecker replied:
Good find. I've moved these to be local to `Read`.

fspmarshall replied:
I can't claim to be an expert on these things, but moving a 64kb buffer into a function-local variable seems iffy to me.  It can be easy to accidentally introduce something that causes go to decide to do a heap allocation for the buffer.  And even if we're not doing a heap allocation for the buffer, it still means go needs to ensure and zero 64kb of stack space each time `Read()` is called (potentially tens of thousands of times per second).

I think we should stick to reusing a persistent buffer for the lifetime of the reader.

dustinspecker replied:
That sounds good me and definitely not wanting to impact performance with this changed. I moved `sizeBytes` as well and `messageBytes` back to fields on the `Reader` struct. They are still unexported, so easy to move in the future if something changes.

zmb3 commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,298 @@
+
+				if n != 0 {
+					// log the number of bytes that were skipped
+					slog.DebugContext(ctx, "skipped dangling data in session recording section", "length", n)
```
I wonder if skipped byes is something that should be tracked in `ReaderStats`.

dustinspecker replied:
I like the idea. I've added a `SkippedBytes` to `ReaderStats`.

zmb3 reviewed (commented):
Some additional suggestions for cleanup/improvements.

We should be cautious here, and I'm fine if you want to decline to apply these, or even push them to a future PR so that this PR is more of a direct migration.

dustinspecker commented:
> Some additional suggestions for cleanup/improvements.
> 
> We should be cautious here, and I'm fine if you want to decline to apply these, or even push them to a future PR so that this PR is more of a direct migration.

Normally, I would lean towards doing a direct migration. But in this case where we're making this importable then I appreciate the suggestions/improvements to make this easier for us to support consumers (plus easier to change the API now before releasing...). Thank you again for the really thorough reviews!

dustinspecker commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,316 @@
+
+// Reader reads Teleport's session recordings
+type Reader struct {
+	inner      io.Reader
```
choose a better name than inner. This holds the limit reader, which is only needed across read iterations to help catch dangling padding bytes.

zmb3 commented on api/sessionrecording/session_recording.go:
```diff
@@ -50,6 +51,8 @@ const (
 
 // NewReader returns a new session recording reader
+//
+// It is the caller's responsibility to call Close on the [Reader] when done.
```
This sentence can be removed, as it doesn't make sense now that we're taking an `io.Reader`. We haven't specified that the input even requires a `Close` method, so of course its the caller's responsibility to do that if necessary.

dustinspecker replied:
I'll clarify this sentence. This is instructing the user they need to call `Close` on this returned `Reader` (`sessionrecording.Reader`), not the provided `io.Reader`.

dustinspecker replied:
Let me know your thoughts on the updated verbiage. Trying to avoid mentioning the `gzip.Reader`, but I can add that if that helps make the message clearer. I'm hoping the updated message is clearer that it's talking about `sessionrecording.Reader`, not an `io.Reader`. Sorry for the confusion.

zmb3 replied:
My mistake. LGTM.

zmb3 commented:
1. We should be able to remove `type bufferCloser struct {` now too.
2. It doesn't look like the Reader's `Reset` method is ever called, so I would remove it to keep the public API as small as possible.
3. Lastly, maybe we can clear up what all these different readers are with some better names and comments:

- rename `reader` to `rawReader`
- rename `inner` to `partReader`

```
// Reader reads Teleport's session recordings
type Reader struct {
	// rawReader is the raw data source we read from
	rawReader io.Reader

	// partReader wraps rawReader and is limited to reading a single
	// (compressed) part from the session recording
	partReader io.Reader

	// gzipReader wraps partReader and decompresses a single part
	// from the session recording
	gzipReader *gzip.Reader

	...
}
```

dustinspecker commented:
> 1. We should be able to remove `type bufferCloser struct {` now too.
> 2. It doesn't look like the Reader's `Reset` method is ever called, so I would remove it to keep the public API as small as possible.
> 3. Lastly, maybe we can clear up what all these different readers are with some better names and comments:
> 
> * rename `reader` to `rawReader`
> * rename `inner` to `partReader`
> 
> ```
> // Reader reads Teleport's session recordings
> type Reader struct {
> 	// rawReader is the raw data source we read from
> 	rawReader io.Reader
> 
> 	// partReader wraps rawReader and is limited to reading a single
> 	// (compressed) part from the session recording
> 	partReader io.Reader
> 
> 	// gzipReader wraps partReader and decompresses a single part
> 	// from the session recording
> 	gzipReader *gzip.Reader
> 
> 	...
> }
> ```

1. Thank you - I completely overlooked this. I updated `newGzipWriter` to take an `io.Writer`. I'm still leaving our `gzipWriter` struct since it interacts with the `writerPool`.
2. I like what you're saying and I'll look at this more. It wasn't as straight forward to remove `Reset`. We have a number of tests that start failing even though it looks like source code isn't using `Reset` anywhere. I'll investigate more, but looks like we _might_ have test code testing test code. I want to understand the tests better before just deleting them...
3. Thank you for the name suggestions! Updated and added godoc.

zmb3 commented on lib/events/filesessions/fileasync_test.go:
```diff
@@ -660,25 +661,52 @@ func emitStream(ctx context.Context, t *testing.T, streamer events.Streamer, inE
+
+	// sort audit events by index
+	slices.SortStableFunc(outEvents, func(a apievents.AuditEvent, b apievents.AuditEvent) int {
+		return int(a.GetIndex() - b.GetIndex())
```
Are you sure this is safe? If `GetIndex()` returns a uint64 this could overflow an int. 

You probably want to convert to signed integers prior to doing the subtraction.

Also, we probably don't need a stable sort since we don't reuse indices on audit events.

dustinspecker replied:
Dumb mistake on my part 🤦 Thank you. Replaced int64 -> int conversion with a comparison and then simply return an int within the `func`.

---

I need to review these test cases, specifically the `TestUploadResume` cases. These tests started failing after replacing calls to `Reader.Reset` with `NewReader`. It looks like these tests were passing due to relying on `sessionrecording.Reader` storing a `lastIndex` and dropping events with new `io.Reader`s provided by `Reader.Reset`, so that's why I have this logic now within this function to drop events with duplicate indices. Which all feels very wrong to me because it's test cases testing functionality that only exists within test implementation, even when `Reader.Reset` existed because `Reader.Reset` was only invoked within this test function.

zmb3 commented on lib/events/filesessions/fileasync_test.go:
```diff
@@ -660,25 +661,52 @@ func emitStream(ctx context.Context, t *testing.T, streamer events.Streamer, inE
+//
+// returned audit events are guaranteed to have unique indexes by removing
+// any audit events containing a previously seen index
+func uniqueAuditEvents(auditEvents []apievents.AuditEvent) []apievents.AuditEvent {
```
Does [`slices.CompactFunc`](https://pkg.go.dev/slices#CompactFunc) work here?

dustinspecker replied:
That's exactly what I was looking for! Forgot about `Compact`.

zmb3 commented on lib/events/filesessions/fileasync_test.go:
```diff
@@ -659,25 +661,36 @@ func emitStream(ctx context.Context, t *testing.T, streamer events.Streamer, inE
-	return outEvents
+
+	// sort audit events by index
+	slices.SortStableFunc(outEvents, func(a apievents.AuditEvent, b apievents.AuditEvent) int {
```
```suggestion
	slices.SortFunc(outEvents, func(a apievents.AuditEvent, b apievents.AuditEvent) int {
```

dustinspecker replied:
Changed. Still feels wrong that I'm having to implement sort/uniq in a test function to drop events with duplicate indices to make a test case pass. More context in this discussion:  https://github.com/gravitational/teleport/pull/50501#discussion_r1901340534 Will investigate more tomorrow. I might be misunderstanding something.

dustinspecker commented on lib/events/filesessions/fileasync_test.go:
```diff
@@ -659,25 +660,31 @@ func emitStream(ctx context.Context, t *testing.T, streamer events.Streamer, inE
-
-		outEvents = append(outEvents, out...)
+	// combine all uploaded parts to create the session recording content
+	var sessionRecordingContent bytes.Buffer
```
For increased confidence, I created a new pull request right off of master to demonstrate this test change + removing `Reset` still passes: #50875

fspmarshall reviewed (commented):
Can we separate out this PR into two separate PRs?  One that renames/moves but changes *no* meaningful code, and another for the changes?

This logic is extremely performance sensitive and extremely difficult to debug when an issue comes up.  We've put a pretty significant amount of dev hours into tuning and debugging of logic related to event decoding & streaming.  We really don't want to accidentally miss a potential issue because it was hidden by the diff of the move, and if a problem does sneak in it will be important to be able to isolate it in the commit log.

dustinspecker commented:
> Can we separate out this PR into two separate PRs? One that renames/moves but changes _no_ meaningful code, and another for the changes?
> 
> This logic is extremely performance sensitive and extremely difficult to debug when an issue comes up. We've put a pretty significant amount of dev hours into tuning and debugging of logic related to event decoding & streaming. We really don't want to accidentally miss a potential issue because it was hidden by the diff of the move, and if a problem does sneak in it will be important to be able to isolate it in the commit log.

Yep, I'll split this out.

dustinspecker commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,309 @@
+/*
+Copyright 2024 Gravitational, Inc.
```
Thank you - updated in the two new Go files in `api/sessionrecording`.

dustinspecker commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,309 @@
+		slog.Int64("skipped-bytes", p.SkippedBytes),
+		slog.Int64("skipped-events", p.SkippedEvents),
+		slog.Int64("out-of-order-events", p.OutOfOrderEvents),
+		slog.Int64("total-events", p.TotalEvents),
```
Hmm. That is interesting about the linter. I'll look around to see if missing some configuration/sloglint version. Otherwise, I'll file an issue upstream.

dustinspecker replied:
Looks similar to this issue: https://github.com/go-simpler/sloglint/issues/37 (https://github.com/go-simpler/sloglint/issues/35)

dustinspecker commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,309 @@
+	// and allows ctx timeout to kick in if specified
+	var checkpointIteration int64
+	for {
+		checkpointIteration++
```
Thank you - updated

dustinspecker commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,309 @@
+			}
+			// read size of this gzipped part as encoded by V1 protocol version
+			_, err = io.ReadFull(r.rawReader, sizeBytes[:Int64Size])
+			if err != nil {
```
Used all of these suggestions for reducing scope of `err` - thank you!

dustinspecker commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,309 @@
+			if err != nil {
+				return nil, r.setError(trace.ConvertSystemError(err))
+			}
+			oneof := apievents.OneOf{}
```
updated

dustinspecker committed:
DNM(api/sessionrecording): copy lib/events for ProtoReader to api

This commit is broken, does not pass tests, and does not even compile.
This commit only exists to make diffs clearer as to what was changed.

The following commands all show no diff:

diff -y api/sessionrecording/sessionlog.go lib/events/sessionlog.go
diff -y api/sessionrecording/stream.go lib/events/stream.go
diff -y api/sessionrecording/stream_test.go lib/events/stream_test.go
diff -y api/sessionrecording/testdata/corrupted-session \
lib/events/testdata/corrupted-session

dustinspecker commented on api/sessionrecording/session_recording.go:
```diff
@@ -0,0 +1,309 @@
+
+// Close releases reader resources
+func (r *Reader) Close() error {
+	if r.gzipReader != nil {
+		return r.gzipReader.Close()
+	}
```
I think you're right that is possible. I'm going to move this conversation to the other pull request that is making implementation changes: #50883
